src/asymmetric_encryption.cpp

```diff
@@ -0,0 +1,242 @@
+                 * '_EVP_PKEY_CTX_set_rsa_oaep_label()' function*/
+                label_copy.reset(static_cast<uint8_t*>(
+                                            _OPENSSL_malloc(oaepPaddingMode.getLabel().size())));
+                memcpy(label_copy.get(),
+                                &oaepPaddingMode.getLabel()[0], oaepPaddingMode.getLabel().size());
+
+                _EVP_PKEY_CTX_set_rsa_oaep_label(keyCtx.get(),
+                                             static_cast<unsigned char*>(label_copy.get()),
+                                             static_cast<int>(oaepPaddingMode.getLabel().size()));
+
+                /* Release ownership from the unique_ptr since the function above takes ownership of
+                 * the label pointer unless it throws an exception. Unique_ptr should only keep
+                 * ownership if an exception is thrown by _EVP_PKEY_CTX_set_rsa_oaep_label.*/
+                label_copy.release();
+            }
+        }
+
+        /* First call to determine the buffer length */
+        _EVP_PKEY_encrypt(keyCtx.get(),
+                          nullptr,
+                          &encryptedMessageLen,
+                          reinterpret_cast<const unsigned char *>(&message.toByteArray()[0]),
+                          message.toByteArray().size());
+
+        /* Allocate memory for the buffer, based on the size returned by _EVP_PKEY_encrypt */
+        encryptedMessage.reset(static_cast<unsigned char*>(
+                                        _OPENSSL_malloc(static_cast<int>(encryptedMessageLen))));
+
+        /* Second call to perform the actual encryption */
+        _EVP_PKEY_encrypt(keyCtx.get(),
+                          encryptedMessage.get(),
+                          &encryptedMessageLen,
+                          reinterpret_cast<const unsigned char *>(&message.toByteArray()[0]),
+                          message.toByteArray().size());
+
+    } catch (const OpenSSLException &e) {
+        throw MoCOCrWException(e.what());
+    }
+
+    return std::vector<uint8_t>(encryptedMessage.get(), encryptedMessage.get() + encryptedMessageLen);
+}
+
+AsymmetricEncryption::CryptoData AsymmetricEncryption::decrypt(AsymmetricPrivateKey key,
+                                                               const RSAPadding &pad,
+                                                               const std::vector<uint8_t>& message)
+{
+    size_t decryptedMessageLen{0};
+    SSL_RSA_OAEP_LABEL_Ptr label_copy{nullptr};
+    SSL_RSA_ENCRYPTION_DATA_Ptr decryptedMessage{nullptr};
+
+    try {
+        auto keyCtx = _EVP_PKEY_CTX_new(key.internal());
+        const auto paddingMode = pad.getPadding();
+
+        _CRYPTO_malloc_init();
+
+        _EVP_PKEY_decrypt_init(keyCtx.get());
+        _EVP_PKEY_CTX_set_rsa_padding(keyCtx.get(), static_cast<int>(paddingMode));
+
+        if(openssl::RSAPaddingMode::OAEP == paddingMode) {
```

Comment:
duplication from `encrypt()`? See my proposal to refactor it into a method of  RSAPadding`.